File: python/Q572_subtree_of_another_tree.py
#
# @lc app=leetcode id=572 lang=python3
#
# [572] Subtree of Another Tree
#


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val: int = val
        self.left: TreeNode | None = left
        self.right: TreeNode | None = right


# @lc code=start
class Solution:
    def same_tree(self, root: TreeNode | None, other: TreeNode | None) -> bool:
        if not root and not other:
            return True

        if not root or not other or root.val != other.val:
            return False

        left_match: bool = self.same_tree(root.left, other.left)
        right_match: bool = self.same_tree(root.right, other.right)
        return left_match and right_match

    def isSubtree(self, root: TreeNode | None, subRoot: TreeNode | None) -> bool:
        if root is None:
            return False

        if self.same_tree(root, subRoot):
            return True

        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)


# @lc code=end

File: rust/src/leetcode/Q572_subtree_of_another_tree.rs
/*
 * @lc app=leetcode id=572 lang=rust
 *
 * [572] Subtree of Another Tree
 */

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @lc code=start
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    fn is_same_tree(p: Option<Rc<RefCell<TreeNode>>>, q: Option<Rc<RefCell<TreeNode>>>) -> bool {
        match (p, q) {
            (None, None) => true,
            (Some(node1), Some(node2)) => {
                if node1.borrow().val == node2.borrow().val {
                    Self::is_same_tree(
                        node1.borrow_mut().left.clone(),
                        node2.borrow_mut().left.clone(),
                    ) && Self::is_same_tree(
                        node1.borrow_mut().right.clone(),
                        node2.borrow_mut().right.clone(),
                    )
                } else {
                    false
                }
            }
            _ => false,
        }
    }

    pub fn is_subtree(
        root: Option<Rc<RefCell<TreeNode>>>,
        sub_root: Option<Rc<RefCell<TreeNode>>>,
    ) -> bool {
        if Self::is_same_tree(root.clone(), sub_root.clone()) {
            true
        } else if let Some(node) = root {
            Self::is_subtree(node.borrow_mut().left.clone(), sub_root.clone())
                || Self::is_subtree(node.borrow_mut().right.clone(), sub_root.clone())
        } else {
            false
        }
    }
}
// @lc code=end
